lib/nat_rewrite.ml

```diff
@@ -11,55 +11,81 @@ end = struct
-    | None ->
-      Lwt.return Untranslated (* don't autocreate new entries *)
-    | Some (_expiry, ((V4 new_src, new_sport), (V4 new_dst, new_dport))) ->
-      (* TODO: we should probably refuse to pass TTL = 0 and instead send an
-          ICMP message back to the sender *)
-      match Nat_decompose.rewrite_packet packet ~src:(new_src, new_sport) ~dst:(new_dst, new_dport) with
-      | Ok packet -> Lwt.return (Translated packet)
-      | Error s -> Log.warn (fun f -> f "Translating a packet failed: %s; packet content: %a" s Nat_packet.pp packet);
-        Lwt.return Untranslated
+    match ports_of_ip transport with
+    | None -> Lwt.return (Error `Untranslated)
+    | Some (sport, dport) -> 
+      let (>>=) = Lwt.bind in
+      (* got everything; do the lookup *)
+      N.lookup table proto ((V4 ip.Ipv4_packet.src), sport) ((V4 ip.Ipv4_packet.dst), dport) >>= function
+      | Some (_expiry, ((V4 new_src, new_sport), (V4 new_dst, new_dport))) ->
+        (* TODO: we should probably refuse to pass TTL = 0 and instead send an
+            ICMP message back to the sender *)
+        begin match Nat_decompose.rewrite_packet packet ~src:(new_src, new_sport) ~dst:(new_dst, new_dport) with
+        | Ok packet -> Lwt.return (Ok packet)
+        | Error s -> Log.warn (fun f -> f "Translating a packet failed: %s; packet content: %a" s Nat_packet.pp packet);
+          Lwt.return (Error `Untranslated)
+        end
+      | _ ->
+        Lwt.return (Error `Untranslated) (* don't autocreate new entries *)
 
   let add_entry mode table packet
       (other_xl_ip, other_xl_port)
       (final_destination_ip, final_destination_port) =
     let `IPv4 (ip_header, transport) = packet in
-    match transport with
-    | `TCP ({Tcp.Tcp_packet.src_port; dst_port; _}, _)
-    | `UDP ({Udp_packet.src_port; dst_port; _}, _) ->
-      let check_scope ip =
-        match Ipaddr.scope ip with
-        | Global | Organization -> true
-        | _ -> false
-      in
-      let (src, dst) = Ipv4_packet.(V4 ip_header.src, V4 ip_header.dst) in
-      match check_scope src, check_scope dst with
-      | false, _ | _, false -> Lwt.return Unparseable
-      | true, true ->
-        let (>>=) = Lwt.bind in
+    let check_scope ip =
+      match Ipaddr.scope ip with
+      | Global | Organization -> true
+      | _ -> false
+    in
+    let (src, dst) = Ipv4_packet.(V4 ip_header.src, V4 ip_header.dst) in
+    let (>>=) = Lwt.bind in
+    match check_scope src, check_scope dst with
+    | false, _ | _, false -> Lwt.return (Error `Cannot_NAT)
+    | true, true ->
+      match transport with
+      | `ICMP _ when mode = Redirect -> Lwt.return (Error (`Cannot_NAT))
+      | `ICMP (icmp, payload) ->
+        begin
+          match ports_of_ip transport with
+          | None -> Lwt.return (Error `Cannot_NAT)
+          | Some (id, _) ->
+            let open Nat_translations in
+            let entries =
+              map_redirect
```

Comment:
Let's see if we can come up with something nicer to solve this.  I appreciate that it's nice to add the ICMP feature, but I would like to make the logic less convoluted rather than more if possible.